### ISSUE TITLE
Preview: get "parent" post ID for non-hierarchical posts when using `asPreview`

### ISSUE THREAD

**Issue by renatonascalves** [wp-graphql/wp-graphql#2876]
### What problem does this address?

Currently, if you try to preview a non-hierarchical post, the `Previewable` interface does not return the post ID of the "parent" post.

<img width="2152" alt="Screenshot 2023-08-01 at 11 22 50" src="https://github.com/wp-graphql/wp-graphql/assets/19148962/da3e33bc-2a16-47f8-a018-879386856b7a">

### What is your proposed solution?

There are a few options here:

- We could add a new field for the `Previewable` interface, exposing the "parent" post ID;
- We could add a new custom interface for non-hierarchical posts exposing the "parent" post ID;
- We could let API consumers to figure this out by themselves.

### What alternatives have you considered?

Currently, API consumers need to circumvent this by getting the id in the params and swap for the one in the response in their apps.

But it is my understanding tools are not doing that and exposing a broken experience. (See the additional context below).

### Additional Context

This is particularly helpful in a toolbar. See this issue for an example.

**Comment by jasonbahl** [comment]
@renatonascalves you have the id in the input already, eh?

**Comment by renatonascalves** [comment]
@jasonbahl I do! And keeping as is and leaving API consumers to handle it in their apps is a valid approach as well. 

I just think the current approach is not obvious. And the example shared shows people has expectations it will return the right `databaseId` when it doesn't.

**Comment by jasonbahl** [comment]
@renatonascalves I think perhaps the solution is maybe clients changing from querying `asPreview: true` to instead querying the node they want, and querying the connected preview node: 

![CleanShot 2023-08-02 at 14 00 11](https://github.com/wp-graphql/wp-graphql/assets/1260765/ec0b615b-041d-4bcb-9a0e-ed569eafb553)

This way you pass the ID of something, you get that something, then you specify you want the connected preview of it. It's a bit more explicit that the client wants 2 nodes, the published node and the preview of the node. 

I believe the `asPreview` is working as intended. The `asPreview` argument allows the consumer to explicitly ask for the preview node, and that node is returned. 

Perhaps in the same way we have a `preview` connection, we could introduce an inverse of that. Something like `previewOf` connection. 

That way we could query the node that a preview is the preview of. 

For example: 


```graphql
{
  post(id: 3192, idType: DATABASE_ID) {
    ...PostTemplateData
    preview { ## The preview node (contains unpublished data )
      node {
        ...PostTemplateData
        previewOf { ## the origin node the preview is a preview of
          node {
            ...PostTemplateData
          }
        }
      }
    }
  }
}

fragment PostTemplateData on Post {
  id
  databaseId
  title
  content
  isPreview
}
```

![CleanShot 2023-08-02 at 14 18 53](https://github.com/wp-graphql/wp-graphql/assets/1260765/eb6cf049-f69c-4341-9f92-2c3a3e4a131b)

Here's the code to register the connection if you want to play with it and see what you think.

```php
add_action( 'graphql_register_types', function() {

	$post_types = WPGraphQL::get_allowed_post_types( 'objects' );

	foreach ( $post_types as $post_type ) {

		// only define the connection for public/publicly_queryable post types
		if ( true !== $post_type->publicly_queryable && true !== $post_type->public ) {
			continue;
		}

		register_graphql_connection( [
			'fromType' => ucfirst( $post_type->graphql_single_name ),
			'toType' => ucfirst( $post_type->graphql_single_name ),
			'description' => __( 'The published node the revision is associated with', 'wp-graphql' ),
			'fromFieldName' => 'previewOf',
			'oneToOne' => true,
			'resolve' => function( \WPGraphQL\Model\Post $preview, $args, $context, $info ) {

				// if the current node is not a preview, return null.
				// this connection is intended to return the published node
				// of the preview
				if ( ! $preview->isPreview ) {
					return null;
				}

				$resolver = new \WPGraphQL\Data\Connection\PostObjectConnectionResolver( $preview, $args, $context, $info );

				// resolve the connection using the preview's parentDatabaseId
				$resolver->set_query_arg( 'p', $preview->parentDatabaseId );

				// return the connection as a one to one connection
				return $resolver->one_to_one()->get_connection();

			}
		] );

	}

} );
```

Let me know what you think and we can include a connection like this in the core WPGraphQL Schema!

**Comment by jasonbahl** [comment]
Here is the same `previewOf` connection simplified with an `asPreview` query:

![CleanShot 2023-08-02 at 14 26 50](https://github.com/wp-graphql/wp-graphql/assets/1260765/2f1dd148-402d-48ad-8a05-3d25bce6a75a)

**Comment by renatonascalves** [comment]
I'll play around with it. But a quick look tells me that a `previewOf` connection seems ideal.

I do wonder however what can be done to avoid this _mistake_ from consumers. It was not obvious to me at first, and others apparently, that using `asPreview` returns a different object.

> I believe the asPreview is working as intended. The asPreview argument allows the consumer to explicitly ask for the preview node, and that node is returned.

And I agree! It is not a bug _per se_. The issue is the change in the response is not obvious.

**Comment by jasonbahl** [comment]
> I do wonder however what can be done to avoid this mistake from consumers. It was not obvious to me at first, and others apparently, that using asPreview returns a different object.

Ya, perhaps more clarity on the description of the `asPreview` argument. 

Right now it says: `Whether to return the node as a preview instance`

Perhaps that can be updated to be more explicit on how the argument will impact resolution. Something like: 

`Whether to return the Preview Node instead of the Published Node. When the ID of a Node is provided along with asPreview being set to true, the preview node with un-published changes will be returned instead of the published node. If no preview node exists or the requestor doesn't have proper capabilities to preview, no node will be returned.`

**Comment by renatonascalves** [comment]
@jasonbahl Tks for improving the description of the argument. 💪🏾 

I liked the suggestios here. And it'd be great to haven them in the core plugin.

**Comment by jeremyProwseYS** [comment]
Just chiming in on this off the back of a related ticket #3001 I raised the other day. In short here are the thoughts I had regarding the current behaviour of the `parent` key when viewing content in preview mode:

I've managed to find a work-around using the `revisionOf` prop to traverse the true parent structure for the post. However this is just that, a "workaround" and I was wondering if there were any plans to reintroduce expected behaviour for the parent prop when preview is concerned. Surely the previewOf prop suggested by @jasonbahl in this post would make more sense to house the info currently returned in the parent prop? I'm just not sure changing the return value / behaviour of a prop based on a variable, such as preview, makes logical sense.

`revisionOf` could be used in place of adding a new prop (previewOf), as it seems to house all the info you would expect.

**Comment by renatonascalves** [comment]
> I'm just not sure changing the return value / behaviour of a prop based on a variable, such as preview, makes logical sense.

So, in a nutshell, WordPress saves revision post types in different post ids. So when a user uses the `asPreview`, aware of it or not, the user is requesting for a different post, not the "main" post, but the revision post.

So having access to the "parent" post via `previewOf` or `revisionOf` makes sense to me. I think it comes down to intention. The intention of using `asPreview` is to grab the revision post, not the main post. So it doesn't make sense to return the main post when trying to preview it.

**Comment by justlevine** [comment]
My understanding of the underlying source of contention is that we don't have a `PostRevision` type, we have a `Post` type, which when using the `asPreview` argument _coerces the revision into a `Post`_.

With that in mind, the discrepancy becomes a lot more clear:

![image](https://github.com/wp-graphql/wp-graphql/assets/29322304/3e65aaa5-654d-4510-a1f8-d46937c45168)

The `parentDatabaseId` is intended to return the parent of a **Post** (aka a `HierarchicalNode`), not of a revision, while the `revisionOf` is meant to contain the revision info.

As such, IMO the **bug** is that when previewed, `HierarchicalNode.parent` (in the screenshot: `previewPage.parentDatabaseId`) refers to the Post this is a revision of when it _should_ return the same value as the `livePage`, and revision traversal should be limited to the `Previewable` and `NodeWithRevisions` interfaces, instead of coopting the semantics.

(I do see the benefit of adding `NodeWithRevisions.revisionOfDatabaseId` to add with that traversal)

**Comment by jasonbahl** [comment]
After working through a request-level preview approach in #3969, I think this is actually solved, and I am happily recanting my earlier stance in this thread.

The root of the problem is that the `asPreview: true` argument swaps the whole node to the revision, so `databaseId` comes back as the revision's id and there is no clean way to get the published post id for a non-hierarchical post.

#3969 adds a `preview` envelope in the request `extensions` that mirrors how WordPress core handles previews. Core keeps the published identity (the preview URL is `?preview_id=43` and the post is still `postid-43`) and overlays the content and featured image from the revision. The new extension does the same: it preserves the node's published identity and overlays only the previewable fields. So `databaseId` stays the published post's id, which is exactly the value the original request here was after, available directly without mapping the revision id back.

In other words, if you switch from `asPreview: true` to the `extensions.preview` envelope, you get the published id on `databaseId` while still seeing the draft content. There is a regression test for this exact case in the PR.

I had previously leaned toward adding a new field or interface to expose the parent id. After looking more closely at how core preserves the parent id rather than overriding it with the preview id, the identity-preserving overlay is the better fit and makes the extra field unnecessary.

The deprecated `asPreview` argument keeps working for now, and the docs include a short upgrade guide.

**Comment by renatonascalves** [comment]
Thanks! 💯
